crates/bevy_reflect/src/kind.rs

```diff
@@ -0,0 +1,307 @@
+                    Self::Tuple(_) => ReflectKind::Tuple,
+                    Self::List(_) => ReflectKind::List,
+                    Self::Array(_) => ReflectKind::Array,
+                    Self::Map(_) => ReflectKind::Map,
+                    Self::Set(_) => ReflectKind::Set,
+                    Self::Enum(_) => ReflectKind::Enum,
+                    Self::Value(_) => ReflectKind::Value,
+                }
+            }
+        }
+
+        impl From<$name$(<$lifetime>)?> for ReflectKind {
+            fn from(value: $name) -> Self {
+                match value {
+                    $name::Struct(_) => Self::Struct,
+                    $name::TupleStruct(_) => Self::TupleStruct,
+                    $name::Tuple(_) => Self::Tuple,
+                    $name::List(_) => Self::List,
+                    $name::Array(_) => Self::Array,
+                    $name::Map(_) => Self::Map,
+                    $name::Set(_) => Self::Set,
+                    $name::Enum(_) => Self::Enum,
+                    $name::Value(_) => Self::Value,
+                }
+            }
+        }
+    };
+}
+
+/// Caused when a type was expected to be of a certain [kind], but was not.
+///
+/// [kind]: ReflectKind
+#[derive(Debug, Error)]
+#[error("kind mismatch: expected {expected:?}, received {received:?}")]
+pub struct ReflectKindMismatchError {
+    pub expected: ReflectKind,
+    pub received: ReflectKind,
+}
+
+macro_rules! impl_cast_method {
+    ($name:ident : Value => $retval:ty) => {
+        #[doc = "Attempts a cast to a [`PartialReflect`] trait object."]
+        #[doc = "\n\nReturns an error if `self` is not the [`Self::Value`] variant."]
+        pub fn $name(self) -> Result<$retval, ReflectKindMismatchError> {
+            match self {
+                Self::Value(value) => Ok(value),
+                _ => Err(ReflectKindMismatchError {
+                    expected: ReflectKind::Value,
+                    received: self.kind(),
+                }),
+            }
+        }
+    };
+    ($name:ident : $kind:ident => $retval:ty) => {
+        #[doc = concat!("Attempts a cast to a [`", stringify!($kind), "`] trait object.")]
+        #[doc = concat!("\n\nReturns an error if `self` is not the [`Self::", stringify!($kind), "`] variant.")]
+        pub fn $name(self) -> Result<$retval, ReflectKindMismatchError> {
+            match self {
+                Self::$kind(value) => Ok(value),
+                _ => Err(ReflectKindMismatchError {
+                    expected: ReflectKind::$kind,
+                    received: self.kind(),
+                }),
+            }
+        }
+    };
+}
+
+/// An immutable enumeration of ["kinds"] of a reflected type.
+///
+/// Each variant contains a trait object with methods specific to a kind of
+/// type.
+///
+/// A [`ReflectRef`] is obtained via [`PartialReflect::reflect_ref`].
+///
+/// ["kinds"]: ReflectKind
+pub enum ReflectRef<'a> {
```

Comment:
Do we want a `non_exhaustive` annotation here? I think probably yes: `Function` at least is liable to be added.

Reply:
I was wondering the same. I chose not to since I'm hoping `Function` gets added in this cycle, and I don't suspect we'll be consistently adding to this list.

But if we think it would be better for future compatibility in case we do touch it, I'm happy to add it!

Reply:
Nah, don't worry about it. Bevy's currently in the "only breaking changes" part of development, so non_exhaustive frankly isn't very useful yet.